doc/Hammer-Use/Hierarchical.rst

```diff
@@ -1,2 +1,110 @@
+
+Tips for Constraining Hierarchical Modules
+------------------------------------------
+
+In a bottom-up hierarchical flow, is is important to remember that submodules do not know the environment in which they will placed. This means:
+
+* At minimum, the pins must be placed on the correct edges of the submodule on metal layers that are accessible in the parent level. Depending on the technology, this may interfere with things like power straps near the edge, so custom obstructions may be necessary. If fixed IOs are placed in submodules (e.g. bumps), then in the parent level, those pins must be promoted up using the ``preplaced: true`` option in the pin assignment.
+
+* Clocks should be constrained more tightly for a submodule compared to its parent to account for extra clock insertion delay, jitter, and skew at increasingly higher levels of hierarchy. Otherwise, you may run into surprise timing violations in submodule instances even if those passed timing in isolation.
+
+* You may need to specify pin delays ``vlsi.inputs.delays`` for many pins to optimize the partitioning of sequential signals that cross the submodule boundary. By default, without pin delay constraints, the input and output delay are constrained to be coincident with its related clock arrival at the module boundary.
+
+* Custom SDC constraints that originate from a higher level (e.g. false paths from async inputs) need to be specified in submodules as well.
+
+* Custom CPFs will need to be written if differently-named power nets need to globally connected between submodules. Similarly, hierarchical flow with custom CPFs can also be used to fake a multi-power domain topology until Hammer properly supports this feature.
+
+Special Notes & Limitations
+---------------------------
+
+#. Currently, Hammer IR keys do not propagate up through the hierarchical tree. For example, if ``vlsi.inputs.clocks`` was specified in the constraints for ``ModuleAA`` but not for ``ModuleAA``, ``ModuleAA`` will not inherit ``ModuleAA``'s constraints; instead, they will take Hammer's defaults.
+
+#. Hammer IR keys specified at the root level (i.e. outside of ``vlsi.inputs.hierarchical.constraints``) are global, i.e. they override submodule constraints unless a meta action is specified.
+
+#. Due to the structure of ``vlsi.inputs.hierarchical.constraints`` as a list structure, currently, there are the following limitations:
+
+    * You must include all of the constraints in a single file. The config parser is unable to combine constraints from differnt files because meta actions do not work on list items. This will make it harder for collaboration, and unfortnately, changes to module constraints at a higher level of hierarchy after submodules are hardened will trigger the Make dependencies, so you will need to modify the generated Makefile or use redo-targets.
```

Comment:
I believe the constraint is more subtle than this.
Meta actions can work on list items, if you want this you need to perform a `deepsusbt`.

Reply:
Sure, `deepsubst` would work, but most others that you would expect to use in non-hierarchical flows wouldn't?